Connector cannot answer for min. 5 minutes if DAPS is unreachable on start-up

# Bug Report

## Describe the Bug
If the DAPS is unreachable on connector start-up it won't be able to answers IDS messages for at least 5 minutes.


### Expected Behavior

- connector should not start and remain in a state where it cannot answer messages, if DAPS is unreachable during start-up
- NullPointerException (in 2nd stack trace) should not be logged with INFO

### Observed Behavior

When the connector is started it tries to connect to the DAPS. If this is not possible an exception is logged, but the connector starts anyway.
Then, if the same connector receives a request it is not able to send out responses and instead answers with an HTTP 500.

## Steps to Reproduce

0. Ensure DAPS is unreachable
1. Start Connector A
2. Ensure DAPS is reachable
3. Start Connector B
4. Request Catalog from Connector A with Connector B
  4.1 Exception on Connector A
  4.2. Connector B receives a 500 response

## Context Information
Every 5 minutes the connector tries to refresh the tokens, hence there is a possibility for the connector to recover.


## Detailed Description

**Exception, logged on Connector start**
```bash
[2022-05-05 13:51:53] [SEVERE ] Error resolving identity provider keys: http://localhost:4567/.well-known/jwks.json
java.net.ConnectException: Failed to connect to localhost/[0:0:0:0:0:0:0:1]:4567
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:297)
	at okhttp3.internal.connection.RealConnection.connect(RealConnection.kt:207)
	at okhttp3.internal.connection.ExchangeFinder.findConnection(ExchangeFinder.kt:226)
	at okhttp3.internal.connection.ExchangeFinder.findHealthyConnection(ExchangeFinder.kt:106)
	at okhttp3.internal.connection.ExchangeFinder.find(ExchangeFinder.kt:74)
	at okhttp3.internal.connection.RealCall.initExchange$okhttp(RealCall.kt:255)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.kt:32)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.kt:95)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.kt:83)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.kt:76)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
	at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
	at okhttp3.internal.connection.RealCall.execute(RealCall.kt:154)
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolver.refreshKeys(IdentityProviderKeyResolver.java:81)
	at org.eclipse.dataspaceconnector.iam.oauth2.core.Oauth2Extension.start(Oauth2Extension.java:119)
	at org.eclipse.dataspaceconnector.boot.system.injection.lifecycle.StartPhase.start(StartPhase.java:34)
	at org.eclipse.dataspaceconnector.boot.system.injection.lifecycle.ExtensionLifecycleManager.start(ExtensionLifecycleManager.java:77)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.eclipse.dataspaceconnector.boot.system.ExtensionLoader.bootServiceExtensions(ExtensionLoader.java:77)
	at org.eclipse.dataspaceconnector.boot.system.runtime.BaseRuntime.bootExtensions(BaseRuntime.java:145)
	at org.eclipse.dataspaceconnector.boot.system.runtime.BaseRuntime.boot(BaseRuntime.java:230)
	at org.eclipse.dataspaceconnector.boot.system.runtime.BaseRuntime.boot(BaseRuntime.java:89)
	at org.eclipse.dataspaceconnector.boot.system.runtime.BaseRuntime.main(BaseRuntime.java:77)
	Suppressed: java.net.ConnectException: Failed to connect to localhost/127.0.0.1:4567
		... 26 more
	Caused by: java.net.ConnectException: Connection refused
		at java.base/sun.nio.ch.Net.pollConnect(Native Method)
		at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
		at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:539)
		at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:594)
		at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
		at java.base/java.net.Socket.connect(Socket.java:633)
		at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
		at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
		... 25 more
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:672)
	at java.base/sun.nio.ch.NioSocketImpl.timedFinishConnect(NioSocketImpl.java:539)
	at java.base/sun.nio.ch.NioSocketImpl.connect(NioSocketImpl.java:594)
	at java.base/java.net.SocksSocketImpl.connect(SocksSocketImpl.java:327)
	at java.base/java.net.Socket.connect(Socket.java:633)
	at okhttp3.internal.platform.Platform.connectSocket(Platform.kt:128)
	at okhttp3.internal.connection.RealConnection.connectSocket(RealConnection.kt:295)
	... 25 more
```


**Exception, logged before 500 response**
```bash
[2022-05-05 13:51:53] [INFO   ] edc-35ebc184-5ef0-4970-a8c0-97075778e06c ready
[2022-05-05 13:52:16] [INFO   ] /api/v1/ids/data
jakarta.servlet.ServletException: java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because the return value of "java.util.concurrent.atomic.AtomicReference.get()" is null
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:410)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:508)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1375)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:463)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1297)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:192)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:562)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because the return value of "java.util.concurrent.atomic.AtomicReference.get()" is null
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolver.resolveKey(IdentityProviderKeyResolver.java:69)
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolver.resolveKey(IdentityProviderKeyResolver.java:46)
	at org.eclipse.dataspaceconnector.common.token.TokenValidationServiceImpl.createVerifier(TokenValidationServiceImpl.java:91)
	at org.eclipse.dataspaceconnector.common.token.TokenValidationServiceImpl.validate(TokenValidationServiceImpl.java:52)
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.Oauth2ServiceImpl.verifyJwtToken(Oauth2ServiceImpl.java:116)
	at org.eclipse.dataspaceconnector.ids.api.multipart.controller.MultipartController.request(MultipartController.java:120)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	... 26 more

[2022-05-05 13:52:16] [INFO   ] unhandled due to prior sendError
jakarta.servlet.ServletException: java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because the return value of "java.util.concurrent.atomic.AtomicReference.get()" is null
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:410)
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:358)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:311)
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:764)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:508)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1375)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:176)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:463)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:174)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1297)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:129)
	at org.eclipse.jetty.server.handler.ContextHandlerCollection.handle(ContextHandlerCollection.java:192)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:122)
	at org.eclipse.jetty.server.Server.handle(Server.java:562)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$0(HttpChannel.java:505)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:762)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:497)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:282)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:319)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:100)
	at org.eclipse.jetty.io.SelectableChannelEndPoint$1.run(SelectableChannelEndPoint.java:53)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:894)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1038)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.NullPointerException: Cannot invoke "java.util.Map.get(Object)" because the return value of "java.util.concurrent.atomic.AtomicReference.get()" is null
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolver.resolveKey(IdentityProviderKeyResolver.java:69)
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolver.resolveKey(IdentityProviderKeyResolver.java:46)
	at org.eclipse.dataspaceconnector.common.token.TokenValidationServiceImpl.createVerifier(TokenValidationServiceImpl.java:91)
	at org.eclipse.dataspaceconnector.common.token.TokenValidationServiceImpl.validate(TokenValidationServiceImpl.java:52)
	at org.eclipse.dataspaceconnector.iam.oauth2.core.identity.Oauth2ServiceImpl.verifyJwtToken(Oauth2ServiceImpl.java:116)
	at org.eclipse.dataspaceconnector.ids.api.multipart.controller.MultipartController.request(MultipartController.java:120)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167)
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$ResponseOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:176)
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:475)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:397)
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:81)
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:255)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248)
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265)
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:234)
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:684)
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394)
	... 26 more
```


## Possible Implementation
-

<sub>Dominik Pinsel <dominik.pinsel@daimler.com>, Mercedes-Benz Tech Innovation GmbH, [legal info/Impressum](https://github.com/Daimler/daimler-foss/blob/master/LEGAL_IMPRINT.md) </sub>

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -57,6 +57,7 @@ in the detailed section referring to by linking pull requests or issues.
 * Fix path conflicts between `CatalogApiController` and `FederatedCatalogApiController` (#1225)
 * Always use configured IDS API path in IDS webhook address (#1249)
 * Fix Azure storage transfer (#1245)
+* Throw exception if `IdentityProviderKeyResolver` cannot get keys at startup (#1266)
 
 ## [milestone-3] - 2022-04-08
 

--- a/extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/Oauth2Extension.java
+++ b/extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/Oauth2Extension.java
@@ -19,6 +19,7 @@ import okhttp3.OkHttpClient;
 import org.eclipse.dataspaceconnector.common.token.TokenGenerationServiceImpl;
 import org.eclipse.dataspaceconnector.common.token.TokenValidationServiceImpl;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolver;
+import org.eclipse.dataspaceconnector.iam.oauth2.core.identity.IdentityProviderKeyResolverConfiguration;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.identity.Oauth2ServiceImpl;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.jwt.DefaultJwtDecorator;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.jwt.Oauth2JwtDecoratorRegistryRegistryImpl;
@@ -37,14 +38,12 @@ import org.eclipse.dataspaceconnector.spi.system.ServiceExtensionContext;
 
 import java.security.PrivateKey;
 import java.security.cert.CertificateEncodingException;
-import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.TimeUnit;
 
 /**
  * Provides OAuth2 client credentials flow support.
  */
-@Provides({IdentityService.class, Oauth2JwtDecoratorRegistry.class, Oauth2ValidationRulesRegistry.class})
+@Provides({ IdentityService.class, Oauth2JwtDecoratorRegistry.class, Oauth2ValidationRulesRegistry.class })
 public class Oauth2Extension implements ServiceExtension {
 
     private static final long TOKEN_EXPIRATION = TimeUnit.MINUTES.toSeconds(5);
@@ -75,10 +74,6 @@ public class Oauth2Extension implements ServiceExtension {
 
     private IdentityProviderKeyResolver providerKeyResolver;
 
-    private long keyRefreshInterval;
-
-    private ScheduledExecutorService executorService;
-
     @Inject
     private OkHttpClient okHttpClient;
 
@@ -90,8 +85,9 @@ public class Oauth2Extension implements ServiceExtension {
     @Override
     public void initialize(ServiceExtensionContext context) {
         var jwksUrl = context.getSetting(PROVIDER_JWKS_URL, "http://localhost/empty_jwks_url");
-        providerKeyResolver = new IdentityProviderKeyResolver(jwksUrl, context.getMonitor(), okHttpClient, context.getTypeManager());
-        keyRefreshInterval = context.getSetting(PROVIDER_JWKS_REFRESH, 5);
+        var keyRefreshInterval = context.getSetting(PROVIDER_JWKS_REFRESH, 5);
+        var identityProviderKeyResolverConfiguration = new IdentityProviderKeyResolverConfiguration(jwksUrl, keyRefreshInterval);
+        providerKeyResolver = new IdentityProviderKeyResolver(context.getMonitor(), okHttpClient, context.getTypeManager(), identityProviderKeyResolverConfiguration);
 
         var configuration = createConfig(context);
 
@@ -116,16 +112,12 @@ public class Oauth2Extension implements ServiceExtension {
 
     @Override
     public void start() {
-        providerKeyResolver.refreshKeys();
-        executorService = Executors.newSingleThreadScheduledExecutor();
-        executorService.scheduleWithFixedDelay(() -> providerKeyResolver.refreshKeys(), keyRefreshInterval, keyRefreshInterval, TimeUnit.MINUTES);
+        providerKeyResolver.start();
     }
 
     @Override
     public void shutdown() {
-        if (executorService != null) {
-            executorService.shutdownNow();
-        }
+        providerKeyResolver.stop();
     }
 
     private static byte[] getEncodedClientCertificate(Oauth2Configuration configuration) {

--- a/extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/IdentityProviderKeyResolver.java
+++ b/extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/IdentityProviderKeyResolver.java
@@ -1,5 +1,5 @@
 /*
- *  Copyright (c) 2020, 2021 Microsoft Corporation
+ *  Copyright (c) 2020 - 2022 Microsoft Corporation
  *
  *  This program and the accompanying materials are made available under the
  *  terms of the Apache License, Version 2.0 which is available at
@@ -9,22 +9,22 @@
  *
  *  Contributors:
  *       Microsoft Corporation - initial API and implementation
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - improvements
  *
  */
 
 package org.eclipse.dataspaceconnector.iam.oauth2.core.identity;
 
-import com.fasterxml.jackson.databind.ObjectMapper;
 import okhttp3.OkHttpClient;
 import okhttp3.Request;
-import okhttp3.Response;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.jwt.JwkKey;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.jwt.JwkKeys;
+import org.eclipse.dataspaceconnector.spi.EdcException;
 import org.eclipse.dataspaceconnector.spi.iam.PublicKeyResolver;
 import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
+import org.eclipse.dataspaceconnector.spi.result.Result;
 import org.eclipse.dataspaceconnector.spi.types.TypeManager;
 
-import java.io.IOException;
 import java.math.BigInteger;
 import java.security.GeneralSecurityException;
 import java.security.KeyFactory;
@@ -35,33 +35,36 @@ import java.util.Base64;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
 import java.util.concurrent.atomic.AtomicReference;
 import java.util.function.Predicate;
 import java.util.stream.Collectors;
 
+import static java.util.Collections.emptyMap;
+import static java.util.concurrent.TimeUnit.MINUTES;
+
 /**
  * Resolves public signing certificates for the identity provider. Used to verify JWTs.
- * This implementation supports key rotation and refresh by calling {@link #refreshKeys()} on a periodic basis.
+ * The keys are cached and the resolver must be started calling the `start` method
  */
-public class IdentityProviderKeyResolver implements PublicKeyResolver, Runnable {
-    private final String jwksUrl;
+public class IdentityProviderKeyResolver implements PublicKeyResolver {
     private final Monitor monitor;
+    private static final String RSA = "RSA";
 
-    private final ObjectMapper mapper;
-    private final AtomicReference<Map<String, RSAPublicKey>> cache = new AtomicReference<>(); // the current key cache, atomic for thread-safety
+    private final TypeManager typeManager;
+    private final IdentityProviderKeyResolverConfiguration configuration;
+    private final ScheduledExecutorService executorService;
+    private final AtomicReference<Map<String, RSAPublicKey>> cache = new AtomicReference<>(emptyMap()); // the current key cache, atomic for thread-safety
     private final OkHttpClient httpClient;
+    private final Predicate<JwkKey> isRsa = key -> RSA.equals(key.getKty());
 
-    /**
-     * Ctor.
-     *
-     * @param jwksUrl     the URL specified by 'jwks_uri' in the document returned by the identity provider's metadata endpoint.
-     * @param typeManager the type manager
-     */
-    public IdentityProviderKeyResolver(String jwksUrl, Monitor monitor, OkHttpClient httpClient, TypeManager typeManager) {
-        this.jwksUrl = jwksUrl;
+    public IdentityProviderKeyResolver(Monitor monitor, OkHttpClient httpClient, TypeManager typeManager, IdentityProviderKeyResolverConfiguration configuration) {
         this.monitor = monitor;
         this.httpClient = httpClient;
-        mapper = typeManager.getMapper();
+        this.typeManager = typeManager;
+        this.configuration = configuration;
+        this.executorService = Executors.newSingleThreadScheduledExecutor();
     }
 
     @Override
@@ -69,63 +72,86 @@ public class IdentityProviderKeyResolver implements PublicKeyResolver, Runnable 
         return cache.get().get(id);
     }
 
-    @Override
-    public void run() {
-        refreshKeys();
-    }
-
-    public void refreshKeys() {
-        try {
-            Request request = new Request.Builder().url(jwksUrl).get().build();
-
-            Response response = httpClient.newCall(request).execute();
-            if (response.code() != 200) {
-                monitor.severe("Unable to refresh identity provider keys. Response code was: " + response.code());
-                return;
-            }
-            try (var body = response.body()) {
-                if (body == null) {
-                    monitor.severe("Unable to refresh identity provider keys. An empty response was returned.");
-                    return;
-                }
-
-                // deserialize the JWKs
-                JwkKeys jwkKeys = mapper.readValue(body.string(), JwkKeys.class);
-                List<JwkKey> keys = jwkKeys.getKeys();
-                if (keys == null || keys.isEmpty()) {
-                    monitor.severe("No keys returned from identity provider.");
-                    return;
-                }
-
-                Map<String, RSAPublicKey> newKeys = deserializeKeys(keys);
-                if (newKeys == null) {
-                    return;
-                }
-
-                cache.set(newKeys);   // reset the cache
-            }
-        } catch (IOException e) {
-            monitor.severe("Error resolving identity provider keys: " + jwksUrl, e);
+    /**
+     * Start the keys cache refreshing job.
+     * Throws exception if it's not able to load the cache at startup.
+     *
+     */
+    public void start() {
+        var result = refreshKeys();
+        if (result.failed()) {
+            throw new EdcException(String.format("Failed to get keys from %s: %s", configuration.getJwksUrl(), String.join(", " + result.getFailureMessages())));
         }
+
+        executorService.scheduleWithFixedDelay(this::refreshKeys, configuration.getKeyRefreshInterval(), configuration.getKeyRefreshInterval(), MINUTES);
     }
 
     /**
-     * Deserializes JWK keys into RSA public keys. Accessible for testing.
+     * Stops the cache refresh job.
      */
-    Map<String, RSAPublicKey> deserializeKeys(List<JwkKey> jwkKeys) {
+    public void stop() {
+        executorService.shutdownNow();
+    }
+
+    /**
+     * Get keys from the JWKS provider. Protected for testing purposes.
+     *
+     * @return succeed if keys are retrieved correctly, failure otherwise
+     */
+    protected Result<Map<String, RSAPublicKey>> getKeys() {
+        try (var response = httpClient.newCall(new Request.Builder().url(configuration.getJwksUrl()).get().build()).execute()) {
+            if (response.code() == 200) {
+                var body = response.body();
+                if (body == null) {
+                    var message = "Unable to refresh identity provider keys. An empty response was returned.";
+                    monitor.severe(message);
+                    return Result.failure(message);
+                }
+
+                var jwsKeys = typeManager.readValue(body.string(), JwkKeys.class);
+                var keys = jwsKeys.getKeys();
+                if (keys == null || keys.isEmpty()) {
+                    var message = "No keys returned from identity provider.";
+                    monitor.warning(message);
+                    return Result.failure(message);
+                }
+
+                return Result.success(deserializeKeys(keys));
+
+            } else {
+                var message = "Unable to refresh identity provider keys. Response code was: " + response.code();
+                monitor.severe(message);
+                return Result.failure(message);
+            }
+        } catch (Exception e) {
+            var message = "Error resolving identity provider keys: " + configuration.getJwksUrl();
+            monitor.severe(message, e);
+            return Result.failure(message);
+        }
+    }
+
+    private Result<Void> refreshKeys() {
+        var result = getKeys();
+        if (result.succeeded()) {
+            cache.set(result.getContent());
+        }
+        return result.map(it -> null);
+    }
+
+    private Map<String, RSAPublicKey> deserializeKeys(List<JwkKey> jwkKeys) {
         return jwkKeys.stream()
-                .filter(new JwkKeyPredicate())
+                .filter(isRsa)
                 .map(this::deserializeKey)
                 .filter(Objects::nonNull)
                 .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
     private Map.Entry<String, RSAPublicKey> deserializeKey(JwkKey key) {
-        BigInteger modulus = unsignedInt(key.getN());
-        BigInteger exponent = unsignedInt(key.getE());
-        RSAPublicKeySpec rsaPublicKeySpec = new RSAPublicKeySpec(modulus, exponent);
+        var modulus = unsignedInt(key.getN());
+        var exponent = unsignedInt(key.getE());
+        var rsaPublicKeySpec = new RSAPublicKeySpec(modulus, exponent);
         try {
-            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
+            var keyFactory = KeyFactory.getInstance(RSA);
             return new AbstractMap.SimpleEntry<>(key.getKid(), (RSAPublicKey) keyFactory.generatePublic(rsaPublicKeySpec));
         } catch (GeneralSecurityException e) {
             monitor.severe("Error parsing identity provider public key, skipping. The kid is: " + key.getKid());
@@ -137,10 +163,4 @@ public class IdentityProviderKeyResolver implements PublicKeyResolver, Runnable 
         return new BigInteger(1, Base64.getUrlDecoder().decode(value));
     }
 
-    static final class JwkKeyPredicate implements Predicate<JwkKey> {
-        @Override
-        public boolean test(JwkKey key) {
-            return "RSA".equals(key.getKty());
-        }
-    }
 }

--- a/extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/IdentityProviderKeyResolverConfiguration.java
+++ b/extensions/iam/oauth2/oauth2-core/src/main/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/IdentityProviderKeyResolverConfiguration.java
@@ -1,0 +1,33 @@
+/*
+ *  Copyright (c) 2022 Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
+ *
+ *  This program and the accompanying materials are made available under the
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - improvements
+ *
+ */
+
+package org.eclipse.dataspaceconnector.iam.oauth2.core.identity;
+
+public class IdentityProviderKeyResolverConfiguration {
+    private final String jwksUrl;
+    private final long keyRefreshInterval;
+
+    public IdentityProviderKeyResolverConfiguration(String jwksUrl, long keyRefreshInterval) {
+        this.jwksUrl = jwksUrl;
+        this.keyRefreshInterval = keyRefreshInterval;
+    }
+
+    public String getJwksUrl() {
+        return jwksUrl;
+    }
+
+    public long getKeyRefreshInterval() {
+        return keyRefreshInterval;
+    }
+}

--- a/extensions/iam/oauth2/oauth2-core/src/test/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/IdentityProviderKeyResolverTest.java
+++ b/extensions/iam/oauth2/oauth2-core/src/test/java/org/eclipse/dataspaceconnector/iam/oauth2/core/identity/IdentityProviderKeyResolverTest.java
@@ -1,5 +1,5 @@
 /*
- *  Copyright (c) 2020, 2021 Microsoft Corporation
+ *  Copyright (c) 2020 - 2022 Microsoft Corporation
  *
  *  This program and the accompanying materials are made available under the
  *  terms of the Apache License, Version 2.0 which is available at
@@ -9,53 +9,114 @@
  *
  *  Contributors:
  *       Microsoft Corporation - initial API and implementation
+ *       Bayerische Motoren Werke Aktiengesellschaft (BMW AG) - improvements
  *
  */
 
 package org.eclipse.dataspaceconnector.iam.oauth2.core.identity;
 
 import com.fasterxml.jackson.databind.ObjectMapper;
+import com.github.javafaker.Faker;
+import okhttp3.Interceptor;
+import okhttp3.MediaType;
 import okhttp3.OkHttpClient;
+import okhttp3.Protocol;
+import okhttp3.Request;
+import okhttp3.Response;
+import okhttp3.ResponseBody;
 import org.eclipse.dataspaceconnector.iam.oauth2.core.jwt.JwkKeys;
 import org.eclipse.dataspaceconnector.spi.EdcException;
 import org.eclipse.dataspaceconnector.spi.monitor.Monitor;
 import org.eclipse.dataspaceconnector.spi.types.TypeManager;
+import org.jetbrains.annotations.NotNull;
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 
 import java.io.IOException;
-import java.io.InputStream;
-import java.security.interfaces.RSAPublicKey;
 import java.util.Map;
 
+import static java.util.Collections.emptyList;
+import static java.util.Collections.emptyMap;
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
 
 class IdentityProviderKeyResolverTest {
-    private static final String JWKS_URL = "https://login.microsoftonline.com/common/discovery/v2.0/keys";
+
+    private static final Faker FAKER = Faker.instance();
+    private static final String JWKS_URL = "https://" + FAKER.internet().url();
     private static final String JWKS_FILE = "jwks_response.json";
+    private final Interceptor interceptor = mock(Interceptor.class);
+    private final TypeManager typeManager = new TypeManager();
+    private final OkHttpClient okHttpClient = new OkHttpClient.Builder().addInterceptor(interceptor).build();
+    private JwkKeys testKeys;
     private IdentityProviderKeyResolver resolver;
-    private JwkKeys keys;
-
-    @Test
-    void verifyJwksDeserialization() {
-        assertThat(keys.getKeys()).isNotNull();
-        Map<String, RSAPublicKey> keyCache = resolver.deserializeKeys(keys.getKeys());
-
-        assertThat(keyCache)
-                .hasSize(5)
-                .containsKey("nOo3ZDrODXEK1jKWhXslHR_KXEg");
-    }
 
     @BeforeEach
     void setUp() {
-        resolver = new IdentityProviderKeyResolver(JWKS_URL, new Monitor() {
-        }, mock(OkHttpClient.class), new TypeManager());
+        resolver = new IdentityProviderKeyResolver(mock(Monitor.class), okHttpClient, typeManager, new IdentityProviderKeyResolverConfiguration(JWKS_URL, 1));
 
-        try (InputStream in = getClass().getClassLoader().getResourceAsStream(JWKS_FILE)) {
-            keys = new ObjectMapper().readValue(in, JwkKeys.class);
+        try (var stream = getClass().getClassLoader().getResourceAsStream(JWKS_FILE)) {
+            testKeys = new ObjectMapper().readValue(stream, JwkKeys.class);
         } catch (IOException e) {
             throw new EdcException("Failed to load keys from file");
         }
+    }
+
+    @Test
+    void getKeys_shouldGetUpdatedKeys() throws IOException {
+        when(interceptor.intercept(any())).thenReturn(response(200, testKeys));
+
+        var result = resolver.getKeys();
+
+        assertThat(result.succeeded()).isTrue();
+        assertThat(result.getContent()).hasSize(5).containsKey("nOo3ZDrODXEK1jKWhXslHR_KXEg");
+    }
+
+    @Test
+    void getKeys_shouldReturnFailureIfServerReturnsError() throws IOException {
+        when(interceptor.intercept(any())).thenReturn(response(500, emptyMap()));
+
+        var result = resolver.getKeys();
+
+        assertThat(result.failed()).isTrue();
+    }
+
+    @Test
+    void getKeys_shouldReturnFailureIfBodyCannotBeDeserialized() throws IOException {
+        when(interceptor.intercept(any())).thenReturn(response(200, null));
+
+        var result = resolver.getKeys();
+
+        assertThat(result.failed()).isTrue();
+    }
+
+    @Test
+    void getKeys_shouldReturnFailureIfNoKeysAreContainedInTheResult() throws IOException {
+        when(interceptor.intercept(any())).thenReturn(response(200, Map.of("keys", emptyList())));
+
+        var result = resolver.getKeys();
+
+        assertThat(result.failed()).isTrue();
+    }
+
+    @Test
+    void start_shouldThrowExceptionIfItFailsToLoadKeysTheFirstTime() throws IOException {
+        when(interceptor.intercept(any())).thenReturn(response(500, emptyMap()));
+
+        assertThatThrownBy(() -> resolver.start()).isInstanceOf(EdcException.class);
+    }
+
+    @NotNull
+    private Response response(int code, Object body) {
+        return new Response.Builder()
+                .code(code)
+                .body(ResponseBody.create(typeManager.writeValueAsString(body), MediaType.get("application/json")))
+                .message(FAKER.twinPeaks().quote())
+                .protocol(Protocol.HTTP_1_1)
+                .request(new Request.Builder().url("http://" + FAKER.internet().url()).build())
+                .build();
     }
 }

